README.markdown
FOP
=============

Installation
------------

FOP should be installed using the [PEAR Installer](http://pear.php.net/). This installer is the backbone of PEAR, which provides a distribution system for PHP packages, and is shipped with every release of PHP since version 4.3.0.

The PEAR channel (`pear.sashabereka.com`) that is used to distribute FOP needs to be registered with the local PEAR environment:

    avb@ubuntu ~ % pear channel-discover pear.sashabereka.com
    Adding Channel "pear.sashabereka.com" succeeded
    Discovery of channel "pear.sashabereka.com" succeeded

This has to be done only once. Now the PEAR Installer can be used to install packages from the channel:

    avb@ubuntu ~ % pear install sashabereka/FOP

After the installation you can find the FOP source files inside your local PEAR directory;

ChangeLog.markdown
FOP 1.0
=================

This is the list of changes for the FOP 1.0 release series.

FOP 1.1.0
-------------------
- added tests
- fixed the issue with not working functionality in case if fopConfXMLRoot was not specified

tests/FOPTest.php
<?php

require_once 'PHPUnit/Autoload.php';
require_once 'FOP.php';

/**
 * Tests for FOP
 *
 * @author    Sasha Bereka <tender.post@gmail.com>
 * @copyright 2011 Sasha Bereka <tender.post@gmail.com>
 * @license   http://www.opensource.org/licenses/bsd-license.php  BSD License
 * @version   Release: 1.1.0
 * @link      http://github.com/sashabereka/FOP/tree
 * @since     Class available since Release 1.1.0
 */


class FOPTest extends PHPUnit_Framework_TestCase {

	private static $_FOP;

	public static function setUpBeforeClass() {
		$config = array();
		$config['templatesRoot'] = realpath(dirname(__FILE__).'/../').'/examples/templates/';
		self::$_FOP = new FOP($config);
	}

	public function testIsInstalled() {
		$this->assertTrue(self::$_FOP->isInstalled());
	}

	public function testGetVersion() {
		$version = self::$_FOP->getVersion();
	}

	public function testRender() {
		$file_path = self::$_FOP
			->setData('var', 'hello world')
			->setTemplateName('example.xsl')
			->render('report')
		;
		$this->assertTrue(file_exists($file_path));
	}

}

tests/FOP/XMLTest.php
<?php

require_once 'PHPUnit/Autoload.php';
require_once 'FOP.php';

/**
 * Tests for FOP_XML
 *
 * @author    Sasha Bereka <tender.post@gmail.com>
 * @copyright 2011 Sasha Bereka <tender.post@gmail.com>
 * @license   http://www.opensource.org/licenses/bsd-license.php  BSD License
 * @version   Release: 1.1.0
 * @link      http://github.com/sashabereka/FOP/tree
 * @since     Class available since Release 1.1.0
 */

class FOP_XMLTest extends PHPUnit_Framework_TestCase {

	/**
	 * @expectedException FOP_Exception
	 */
	public function testSetData() {
		$XML = new FOP_XML();
		$XML->setData('35', 'value');
	}

	public function testSave() {
		$XML = new FOP_XML();
		$XML->setData('var', 'value');
		$filePath = $XML->save();
		$this->assertTrue(file_exists($filePath));
	}

}

tests/FOP/ConfigTest.php
<?php

require_once 'PHPUnit/Autoload.php';
require_once 'FOP.php';

/**
 * Tests for FOP_Config
 *
 * @author    Sasha Bereka <tender.post@gmail.com>
 * @copyright 2011 Sasha Bereka <tender.post@gmail.com>
 * @license   http://www.opensource.org/licenses/bsd-license.php  BSD License
 * @version   Release: 1.1.0
 * @link      http://github.com/sashabereka/FOP/tree
 * @since     Class available since Release 1.1.0
 */

class FOP_ConfigTest extends PHPUnit_Framework_TestCase {

	private $_Config;

	public function setUp() {
		$config = array();
		$config['templatesRoot'] = '/templates/';
		$config['libraryRoot'] = '/path/to/library/';
		$config['debug'] = true;

		$this->_Config = FOP_Config::getInstance();
		$this->_Config->init($config);
	}

	public function testGet() {
		$this->assertTrue($this->_Config->debug);
		$this->assertEquals('downloaded', $this->_Config->downFilename);
	}

	/**
	 * @depends testGet
	 */
	public function testSet() {
		$this->_Config->debug = false;
		$this->assertFalse($this->_Config->debug);
		$this->_Config->templatesRoot = '/path/to/templates/';
		$this->assertEquals('/path/to/templates/', $this->_Config->templatesRoot);
	}

	/**
	 * @depends testGet
	 */
	public function testToArray() {
		$expected = array(
			'libraryRoot'=>'/path/to/library/',
			'templatesRoot'=>'/templates/',
			'tmpRoot'=>'/tmp/',
			'fopConfXMLRoot'=>'',
			'pathToFOP'=>'',
			'downFilename'=>'downloaded',
			'debug'=>true
		);
		$this->assertEquals($expected, $this->_Config->toArray());
	}

	public function tearDown() {}

}

FOP.php
<?php

/**
 * FOP
 *
 * Copyright (c) 2011, Sasha Bereka <tender.post@gmail.com>.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *   * Neither the name of Sasha Bereka nor the names of his
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * @package   FOP
 * @author    Sasha Bereka <tender.post@gmail.com>
 * @copyright 2011 Sasha Bereka <tender.post@gmail.com>
 * @license   http://www.opensource.org/licenses/bsd-license.php  BSD License
 * @since     File available since Release 1.0.0
 */

$current_dir = dirname(__FILE__).'/';
require_once($current_dir.'FOP/Autoload.php');

/**
 * FOP is php wrapper for Apache FOP application
 * Generates pdf files using xsl template
 *
 * @author    Sasha Bereka <tender.post@gmail.com>
 * @copyright 2011 Sasha Bereka <tender.post@gmail.com>
 * @license   http://www.opensource.org/licenses/bsd-license.php  BSD License
 * @version   Release: 1.1.0
 * @link      http://github.com/sashabereka/FOP/tree
 * @since     Class available since Release 1.0.0
 */

class FOP {

    private $_xml;
    private $_xsl;

    public function __construct($config = array()) {
		$this->_initConfig($config);
        $this->_initXML();
        $this->_initXSL();
    }

	private function _initConfig($config = array()) {
		$Config = FOP_Config::getInstance();
		$Config->init($config);
	}

    private function _initXML() {
        $this->_xml = new FOP_XML();
    }

    private function _initXSL() {
        $this->_xsl = new FOP_XSL();
    }

	/**
	 * Checks if Apache FOP application is installed on server
	 * @return bool
	 */
	public function isInstalled() {
		try {
			$this->getVersion();
		} catch(FOP_Exception $e) {
			return false;
		}
		return true;
	}

	/**
	 * Returns Apache FOP application version
	 * @throws FOP_Exception
	 * @return   string
	 */
	public function getVersion() {
		$cmd = 'fop -v 2>&1';
		exec($cmd, $output);
		if (!count($output)) {
			throw new FOP_Exception('command output is empty', 1001);
		}
		foreach ($output as $line) {
			if (preg_match('~FOP Version ([\d\.]+)~', $line, $matches)) {
				return $matches[1];
			}
		}
		throw new FOP_Exception('can\'t get Apache FOP version', 1002);
	}

	/**
	 * Returns configuration parameters
	 * @return  array
	 */
	public function getConfiguration() {
		$Config =  FOP_Config::getInstance();
		return $Config->toArray();
	}

    /**
     * Stores data for pdf template
     * @param string $var
     * @param mixed $value
     * @return FOP
     */
    public function setData($var, $value, $escaping = true) {
        $this->_xml->setData($var, $value, $escaping);
        return $this;
    }

    /**
     * Set xsl template name to use when generating pdf
     * @param string $templateName
     * @return FOP
     */
    public function setTemplateName($template_name) {
        $this->_xsl->setFilename($template_name);
        return $this;
    }

    /**
     * Returns path to rendered pdf file
     * @throws FOP_Exception
     * @return string
     */
    public function render() {
        $Config = FOP_Config::getInstance();
        $pdf_file_path = $Config->tmpRoot.$this->_renderTmpPDFFilename();
        if (!$this->_xsl->isValid()) {
            throw new FOP_Exception('Can\'t find xsl template file: "'.$this->_xsl->getFilePath().'"', 1008);
        }
        $xml_file_path = $this->_xml->save();
        $conf = '';
	    if ($Config->fopConfXMLRoot and is_file($Config->fopConfXMLRoot)) {
	        $conf = '-c '.$Config->fopConfXMLRoot;
        }
        $bash = $Config->pathToFOP.'fop '.$conf.' -xml '.$xml_file_path.' -xsl '.$this->_xsl->getFilePath().' -pdf '.$pdf_file_path.' 2>&1';
        exec($bash, $output);
        unlink($xml_file_path);
	    if (!file_exists($pdf_file_path)) {
			$this->_processOutput($output);
        }
        return $pdf_file_path;
    }

    private function _processOutput($output) {
	    if (!count($output)) {
		    return;
	    }
	    $output_error = "";
		if (FOP_Config::getInstance()->debug) {
			$write_error = false;
			foreach ($output as $line) {
				if (strpos($line, 'Fatal Error') !== false) {
					$write_error = true;
				}
				if (strpos($line, 'SEVERE: Exception') !== false) {
					$write_error = false;
				}
				if ($write_error) {
					$output_error .= $line." \n";
				}
			}
		}
		$exception_message = 'Rendering pdf error';
		if ($output_error) {
			$exception_message .= ":\n".$output_error;
		}
		throw new FOP_Exception($exception_message, 1009);
    }

    /**
     * Render and flush pdf file
     * @throws FOP_Exception
     * @param string $filename
     */
    public function renderAndFlush($filename = '') {
        if (empty($filename)) {
            $filename = FOP_Config::getInstance()->downFilename;
        }
        $this->flush($this->render(), $filename);
    }

	/**
	 * Flush pdf file
	 * @param string  $pdfFilePath
	 * @param string  $filename
	 * @return void
	 */
	public function flush($pdfFilePath, $filename) {
		header("Content-type:application/pdf");
        header("Content-Disposition:attachment;filename=$filename.pdf");
        readfile($pdfFilePath);
        die();
	}

    private function _renderTmpPDFFilename() {
        return 'fop_'.str_replace('.', '', microtime('true')).'.pdf';
    }

}

FOP/XSL.php
<?php

/**
 * FOP
 *
 * Copyright (c) 2011, Sasha Bereka <tender.post@gmail.com>.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *   * Neither the name of Sasha Bereka nor the names of his
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * @package   FOP
 * @author    Sasha Bereka <tender.post@gmail.com>
 * @copyright 2011 Sasha Bereka <tender.post@gmail.com>
 * @license   http://www.opensource.org/licenses/bsd-license.php  BSD License
 * @since     File available since Release 1.0.0
 */

/**
 * FOP_XSL is used for getting and checking FOP input xsl file
 *
 * @author    Sasha Bereka <tender.post@gmail.com>
 * @copyright 2011 Sasha Bereka <tender.post@gmail.com>
 * @license   http://www.opensource.org/licenses/bsd-license.php  BSD License
 * @version   Release: 1.1.0
 * @link      http://github.com/sashabereka/FOP/tree
 * @since     Class available since Release 1.0.0
 */

class FOP_XSL {

    protected $_folder;
    protected $_filename;

    public function __construct() {
        $this->_initFolder();
    }    

    /**
     * @return string
     */
    public function getFolder() {
        return $this->_folder;
    }

    /**
     * @param string $filename
     */
    public function setFilename($filename) {
        $this->_filename = $filename;
    }

    /**
     * @return string
     */
    public function getFilename() {
        return $this->_filename;
    }

    private function _initFolder() {
        $Config = FOP_Config::getInstance();
        $this->_folder = $Config->templatesRoot;
	    if (!is_dir($this->_folder)) {
			throw new FOP_Exception('Can\'t find pdf templates directory "'.$this->_folder.'"', 1005);
	    }
    }

    /**
     * Return file path to xsl template
     * @return string
     */
    public function getFilePath() {
        return $this->getFolder().$this->getFilename();
    }
	
	public function isValid() {
		if (is_file($this->getFilePath())) {
			return true;
		}
		return false;
	}

}

FOP/Config.php
<?php

/**
 * FOP
 *
 * Copyright (c) 2011, Sasha Bereka <tender.post@gmail.com>.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *   * Neither the name of Sasha Bereka nor the names of his
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * @package   FOP
 * @author    Sasha Bereka <tender.post@gmail.com>
 * @copyright 2011 Sasha Bereka <tender.post@gmail.com>
 * @license   http://www.opensource.org/licenses/bsd-license.php  BSD License
 * @since     File available since Release 1.0.0
 */

/**
 * FOP_Config contains default config values
 * default config values should not be changed
 * use FOP::__construct($config) parameter to change default config values
 *
 * @author    Sasha Bereka <tender.post@gmail.com>
 * @copyright 2011 Sasha Bereka <tender.post@gmail.com>
 * @license   http://www.opensource.org/licenses/bsd-license.php  BSD License
 * @version   Release: 1.1.0
 * @link      http://github.com/sashabereka/FOP/tree
 * @since     Class available since Release 1.0.0
 */

class FOP_Config {

    public $libraryRoot;
    /**
     * path to xsl templates
     * @var string
     */
	public $templatesRoot = '';
    /**
     * path to temporary generated pdf files
     * @var string
     */
	public $tmpRoot = '/tmp/';
	/**
	 *path to additional xml config file
	 *specifies location for custom fonts
	 * @var string
	 */
	public $fopConfXMLRoot = '';
    /**
     * path to fop application
     * @var string
     */
	public $pathToFOP = '';
    /**
     * default name of generated pdf file
     * @var string
     */
	public $downFilename = 'downloaded';
	public $debug = false;

    protected static $_instance;

    protected function __construct() {
        $this->libraryRoot = realpath(dirname(__FILE__).'/../').'/';
    }

	protected function __clone() {}

	public static function getInstance() {
		if (!self::$_instance instanceof self) {
			self::$_instance = new self;
		}
		return self::$_instance;
	}

	public function __set($name, $value) {
		if (!isset($this->$name)) {
			throw new FOP_Exception('FOP config property "'.$name.'" doesn\'t exist', 1003);
		}		
		$this->$name = $value;
	}

	public function init($config = array()) {
		if (!is_array($config) or !count($config)) {
			return;
		}
		foreach ($config as $name=>$value) {
			$this->$name = $value;
		}
	}

	public function toArray() {
		return get_object_vars(self::$_instance);
	}

}

FOP/Autoload.php
<?php

/**
 * FOP
 *
 * Copyright (c) 2011, Sasha Bereka <tender.post@gmail.com>.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *   * Neither the name of Sasha Bereka nor the names of his
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * @package   FOP
 * @author    Sasha Bereka <tender.post@gmail.com>
 * @copyright 2011 Sasha Bereka <tender.post@gmail.com>
 * @license   http://www.opensource.org/licenses/bsd-license.php  BSD License
 * @since     File available since Release 1.1.0
 */


function fop_autoload($class) {
	$class_path = str_replace('_', '/', $class);
	$fop_dir = realpath(dirname(__FILE__).'/../').'/';
	$class_path = $fop_dir.$class_path.'.php';
	if (is_file($class_path)) {
		require_once($class_path);
	}
}

spl_autoload_register('fop_autoload');

FOP/Exception.php
<?php

/**
 * FOP
 *
 * Copyright (c) 2011, Sasha Bereka <tender.post@gmail.com>.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *   * Neither the name of Sasha Bereka nor the names of his
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * @package   FOP
 * @author    Sasha Bereka <tender.post@gmail.com>
 * @copyright 2011 Sasha Bereka <tender.post@gmail.com>
 * @license   http://www.opensource.org/licenses/bsd-license.php  BSD License
 * @since     File available since Release 1.0.0
 */

/**
 * Exception codes:
 * 1001 - Command output is empty
 * 1002 - Can't get Apache FOP version
 * 1003 - FOP config property doesn't exist
 * 1004 - Variable name for pdf template is invalid
 * 1005 - Directory for storing xml data file doesn't exist and can't be created
 * 1006 - Directory for storing xml data file is not writable
 * 1007 - Can't find xml data file
 * 1008 - Can't find xsl template file
 * 1009 - Rendering pdf error
 *
 * @author    Sasha Bereka <tender.post@gmail.com>
 * @copyright 2011 Sasha Bereka <tender.post@gmail.com>
 * @license   http://www.opensource.org/licenses/bsd-license.php  BSD License
 * @version   Release: 1.0.0
 * @link      http://github.com/sashabereka/FOP/tree
 * @since     Class available since Release 1.0.0
 */

class FOP_Exception extends Exception {}

FOP/XML.php
<?php

/**
 * FOP
 *
 * Copyright (c) 2011, Sasha Bereka <tender.post@gmail.com>.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *   * Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *   * Neither the name of Sasha Bereka nor the names of his
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * @package   FOP
 * @author    Sasha Bereka <tender.post@gmail.com>
 * @copyright 2011 Sasha Bereka <tender.post@gmail.com>
 * @license   http://www.opensource.org/licenses/bsd-license.php  BSD License
 * @since     File available since Release 1.0.0
 */

/**
 * FOP_XML is used for creating FOP input xml file
 *
 * @author    Sasha Bereka <tender.post@gmail.com>
 * @copyright 2011 Sasha Bereka <tender.post@gmail.com>
 * @license   http://www.opensource.org/licenses/bsd-license.php  BSD License
 * @version   Release: 1.1.0
 * @link      http://github.com/sashabereka/FOP/tree
 * @since     Class available since Release 1.0.0
 */

class FOP_XML {

    protected $_output;
	protected $_folder;
	protected $_filename;

    function __construct() {
        $this->_initFolder();
        $this->_initFilename();
		$this->_output = new DOMDocument("1.0", "ISO-8859-1");
		$rootNode = $this->_output->createElement("document");
		$this->_output->appendChild($rootNode);
        $this->setData('templatesRoot', FOP_Config::getInstance()->templatesRoot);
	}

    /**
     * @param string $tag
     * @param mixed $data
     */
    public function __set($tag, $data) {
		$this->setData($tag, $data);
	}

	/**
	 * Stores data for pdf  template
     * @param string $tag
     * @param mixed $data
     */
    public function setData($tag, $data, $escaping = true) {
	    if (is_object($data)) {
			$this->_assignArray($this->_output->firstChild, $tag, (array) $data, $escaping);
	    } elseif (is_array($data)) {
			$this->_assignArray($this->_output->firstChild, $tag, $data, $escaping);
		} else {
			$this->_assignVar($this->_output->firstChild, $tag, $data, $escaping);
		}
	}

	/**
	 * get filepath of generated xml data file
     * @return string
     */
    public function getFilePath() {
        return $this->getFolder().$this->getFilename();
    }

    public function getFolder() {
        return $this->_folder;
    }

    public function getFilename() {
        return $this->_filename;
    }

	/**
	 * saves stored data into xml file and returns xml file path
	 * @throws FOP_Exception
	 * @return string
	 */
    public function save() {
	    if (!is_writable($this->_folder) and !chmod($this->_folder, 0777)) {
		    throw new FOP_Exception('Directory "'.$this->_folder.'" for storing xml data file is not writable', 1006);
	    }
        $this->_output->save($this->getFilePath());
	    if (!is_file($this->getFilePath())) {
            throw new FOP_Exception('Can\'t find xml data file: "'.$this->getFilePath().'"', 1007);
        }
	    return $this->getFilePath();
	}

	private function _assignVar($parentElement, $tagName, $data, $escaping) {
		$this->_checkTagName($tagName);
		if (is_numeric($data) or !$escaping) {
			$node = $this->_output->createElement($tagName, $data);
		} else {
			$CDATA = $this->_output->createCDATASection($data);
			$node = $this->_output->createElement($tagName);
			$node->appendChild($CDATA);
		}
		$parentElement->appendChild($node);
	}

	private function _assignArray($parentElement, $tagName, $data, $escaping) {
		$this->_checkTagName($tagName);
		$node = $this->_output->createElement($tagName);
		foreach ($data as $index => $value) {
			$tagName = (is_numeric($index))?  'item' : $index;
			if (is_array($value) or is_object($value)) {
				$this->_assignArray($node, $tagName, $value, $escaping);
			} else {
				$this->_assignVar($node, $tagName, $value, $escaping);
			}
		}
		$parentElement->appendChild($node);
	}

	private function _checkTagName($tagName) {
		if (preg_match("/(?:^(?:(?:xml)|(?:[^a-z])))|(?:[<>&'\"\s])/i",$tagName)) {
			throw new FOP_Exception('Variable name "'.$tagName.'" is invalid', 1004);
		}
	}

	/**
	 * in case if directory for storing xml data doesn't exist, attempts to create it
	 * @throws FOP_Exception
	 * @return void
	 */
    private function _initFolder() {
        $this->_folder = FOP_Config::getInstance()->tmpRoot;
	    if (!is_dir($this->_folder) and !mkdir($this->_folder, 0777, true)) {
			throw new FOP_Exception('Directory "'.$this->_folder.'" for storing xml data file doesn\'t exist and can\'t be created', 1005);
	    }
    }

    private function _initFilename() {
        $this->_filename = 'fop_'.str_replace('.', '', microtime('true')).'.xml';
    }

}